### PR TITLE
Allow Edge, Improve authentication

### PR DESCRIPTION
**Pull request by mrmueller**
Allows to launch Microsoft Edge

start-process -wait does not work if browser was already open before start-process. Stop pode server after first request instaed.
Job will go to "Completed" state and remain in that state.

### CODE REVIEW

**Comment by pcgeek86** [comment]
@mrmueller Did you want me to merge this? I'm not seeing an option to do so.

**Comment by mrmueller** [comment]
Found a bug in my code...
Looks like I was not able to improve from what you did ;)

Closed the PR. Will submit again when i figure out a better option...

**Comment by mrmueller** [comment]
Sorry for the spam... I am new to this. Trying to do a pull request for only the functions/Grant-YouTubeOauth.ps1 file...

**Comment by mrmueller** [comment]
I did not mean to delete functions/Get-YouTubeChannel.ps1... can you kick this out somehow?
Edit: Managed to add it pack to the PR. Added the -Raw parameter.

The changes to functions/Grant-YouTubeOauth.ps1 and functions/private/Find-Browser.ps1 should be more helpful.
Allows the script to discover Microsoft Edge.
Pode server will be stopped when the access token is received
With this, the job completes automatically. No need to stop it manually.
Added URL encoding for the scopes. Withouth this, a browser tab was opened for every scope.

**Comment by mrmueller** [comment]
I think this is ready to merge now. Hope you agree :)

**Comment by mrmueller** [comment]
@pcgeek86  do you want to merge this PR?

File: functions/Get-YouTubeChannel.ps1 [hunk 1]
```diff
@@ -22,6 +23,10 @@ function Get-YouTubeChannel {
   $Result = Invoke-RestMethod -Uri $Uri -Method Get -Headers (Get-AccessToken)
   $Result.items | ForEach-Object -Process { $PSItem.PSTypeNames.Add('YouTube.Channel') }
 
-  if ($Raw) { return $Result }
-  $Result.items
+  if ($Raw) {
```
Comment (pcgeek86) [inline functions/Get-YouTubeChannel.ps1]: Why did you change the logic here? It's more concise to simply return `$Result` in an `if` statement. If that parameter isn't passed in, then skip and return the items. No need to put an else statement.

File: functions/Grant-YouTubeOauth.ps1 [hunk 2]
```diff
@@ -43,16 +48,10 @@ function Grant-YoutubeOauth {
     'https://www.googleapis.com/auth/youtube.force-ssl'
     'https://www.googleapis.com/auth/youtube.readonly'
   )
-  $Uri = 'https://accounts.google.com/o/oauth2/v2/auth?include_granted_scopes=true&response_type=token&client_id={0}&redirect_uri={1}&scope={3}&state={2}' -f $Client.client_id, $RedirectUri, (New-Guid).Guid, ($ScopeList -join ' ')
+  $scopesURI = [System.Web.HttpUtility]::UrlEncode($ScopeList -join ' ')
+  $Uri = 'https://accounts.google.com/o/oauth2/v2/auth?include_granted_scopes=true&response_type=token&client_id={0}&redirect_uri={1}&scope={2}&state={2}' -f $Client.client_id, $RedirectUri, ($scopesURI), (New-Guid).Guid
```
Comment (pcgeek86) [inline functions/Grant-YouTubeOauth.ps1:53]: Does this fix something? No need for the parentheses around `($scopesURI)`.

File: functions/Grant-YouTubeOauth.ps1 [hunk 3]
```diff
@@ -43,16 +48,10 @@ function Grant-YoutubeOauth {
...
-  Start-Process -FilePath $Browser -ArgumentList ('"{0}"' -f $Uri) -Wait
-
-  if (!$IsMacOS) {
-    Stop-Job -Name $JobName
-  }
-  else {
-    Write-Warning -Message 'Please manually stop the Background Job running the Pode web server after completing authentication. Cannot wait for browser process to complete on MacOS.'
-  }
-}
+  Start-Process -FilePath $Browser -ArgumentList ('"{0}"' -f $Uri)
```
Comment (pcgeek86) [inline functions/Grant-YouTubeOauth.ps1:56]: I remember putting this logic in for MacOS. Did you test this on MacOS?

File: functions/private/Find-Browser.ps1 [hunk 4]
```diff
@@ -24,6 +24,7 @@ function Find-Browser {
     'firefox.exe'
     'chrome'
     'firefox'
+    'MicrosoftEdge.exe'
   )
```
Comment (pcgeek86) [inline functions/private/Find-Browser.ps1:28]: LGTM

File: functions/Grant-YouTubeOauth.ps1 [hunk 5]
```diff
@@ -17,23 +17,28 @@ function Grant-YoutubeOauth {
...
-
         $Response = @'
         <h1 style="font-family: sans-serif;">Authentication Complete</h1>
         <h3 style="font-family: sans-serif;">You may close this browser window.</h3>
         <script>
           console.log(window.location.hash);
           let access_token_regex = /access_token=(?<access_token>.*?)&token_type/;
           let result = access_token_regex.exec(window.location.hash);
-          fetch(`/auth/complete?access_token=${result.groups.access_token}`);
+          fetch(`/auth/receive?access_token=${result.groups.access_token}`);
```
Comment (pcgeek86) [inline functions/Grant-YouTubeOauth.ps1:27]: What's happening here, with the newly added `/receive` route?

**Comment by pcgeek86** [comment]
@mrmueller Left some comments.

- Adding the `-Raw` parameter is fine, because it was missing. However, the logic in the `Get-YouTubeChannel` function should be fine. No need to change that.
- I don't recall a new browser tab opening for each oAuth scope. Which platforms did you test your new changes on?

Thanks!

File: functions/Grant-YouTubeOauth.ps1 (hunk 3 above)
Comment (mrmueller) [reply functions/Grant-YouTubeOauth.ps1:56]: no longer needed because the job can always complete. see above.
I do not have a MAC. Did not test.

File: functions/Grant-YouTubeOauth.ps1 (hunk 5 above)
Comment (mrmueller) [reply functions/Grant-YouTubeOauth.ps1:27]: Check it out at line 33.

/complete is your solution to parse the token. It was invoking itself after parsing
now it invokes /receive
/receive returns the token and closes the pode server. This allows the Powershell job to complete reliably.

File: functions/Grant-YouTubeOauth.ps1 (hunk 2 above)
Comment (mrmueller) [reply functions/Grant-YouTubeOauth.ps1:53]: without encoding, Edge on Windows opens each scope "URL" in a separate tab.

**Comment by mrmueller** [comment]
@pcgeek86 Replied to comments
Reverted the change in `Get-YouTubeChannel`
Tested only on Win10 with Edge

**Comment by pcgeek86** [comment]
Looks good. Let me test this out on my machine.

Unfortunately I don't have a Mac to test on either. However, your proposed solution to have Pode shut itself down should theoretically be adequate.
